### PR TITLE
tenant: add endpoint with instant metrics

### PR DESCRIPTION
**Pull request by darinpp**
Previously the tenant process was serving various metrics on
`/_status/vars`. This endpoint has all the available metrics and these are
updated every 10 sec. Many of the metrics show a rate that is calculated
over the 10 sec interval. Some of the metrics are used by the cockroach
operator to monitor the CPU workload of the tenant process and use that
workload for automatic scaling. The 10 sec interval however is too long
and causes a slow scaling up. The reporting of high CPU utilization can
take up to 20 sec (to compute a delta). To resolve this, the PR adds a
new endpoint `/_status/load` that provides an instant reading of a
very small subset of the normal metrics - user and system CPU time for
now. By having these be instant, the client can retrieve in quick
succession, consecutive snapshots and compute a precise CPU utulization.
It also allows the client to control the interval between the two pulls
(as opposed to having it hard coded to 10 sec).

Release note: None

### CODE REVIEW

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @darinpp)

___
*[-- commits, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkcvxUa25V-GvMNs5p3):*
NIT: Update the commit/PR message to be `/_status/load`.
___
*[pkg/server/tenant.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkcyAm10MMmjno0S0-d:-MkcyAm10MMmjno0S0-e:bu60xdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L321)):*
> ```Go
> 
> 	return func(w http.ResponseWriter, r *http.Request) {
> 		if err := cpuTime.Get(os.Getpid()); err != nil {
> ```

Should this CPU calculation code be an exported function on `status.RuntimeStatSampler`? Something like:
```
func (rsr *RuntimeStatSampler) SampleCPU() {
  ...
}
```
I believe this code should be thread-safe. You could call that code both from here and from the `SampleEnvironment` method. It would get the updated CPU time and then update the `CPUUserNS` and `CPUSysNS` gauges.
___
*[pkg/server/tenant.go, line 323 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkd-EXg82PTC4GeXrFJ:-Mkd-EXhAM8kbu6UPN5o:b-4118zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L323)):*
> ```Go
> 		if err := cpuTime.Get(os.Getpid()); err != nil {
> 			log.Ops.Errorf(ctx, "unable to get cpu usage: %v", err)
> 			http.Error(w, err.Error(), http.StatusInternalServerError)
> ```

Don't we need ` `return` here in the error code path?
___
*[pkg/server/tenant.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkd-LjsDo_ZAOpyv5Wq:-Mkd-LjsDo_ZAOpyv5Wr:b-77ht6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L334)):*
> ```Go
> 		if err := exporter.PrintAsText(w); err != nil {
> 			log.Errorf(r.Context(), "%v", err)
> 			http.Error(w, err.Error(), http.StatusInternalServerError)
> ```

And a `return` here as well.

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[-- commits, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkcvxUa25V-GvMNs5p3):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Update the commit/PR message to be `/_status/load`.
</blockquote></details>

done
___
*[pkg/server/tenant.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkcyAm10MMmjno0S0-d:-MkdLUw4BulTq5rYVOw7:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L321)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should this CPU calculation code be an exported function on `status.RuntimeStatSampler`? Something like:
```
func (rsr *RuntimeStatSampler) SampleCPU() {
  ...
}
```
I believe this code should be thread-safe. You could call that code both from here and from the `SampleEnvironment` method. It would get the updated CPU time and then update the `CPUUserNS` and `CPUSysNS` gauges.
</blockquote></details>

changed
___
*[pkg/server/tenant.go, line 323 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkd-EXg82PTC4GeXrFJ:-MkdLVu86J2ZrzR_otZ-:b-ro2cm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L323)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Don't we need ` `return` here in the error code path?
</blockquote></details>

after the `SampleCPU` change - not needed.
___
*[pkg/server/tenant.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkd-LjsDo_ZAOpyv5Wq:-MkdLavX5EH8jscDmm30:b-5co7g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L334)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

And a `return` here as well.
</blockquote></details>

after the `SampleCPU` change - not needed.

**Comment by andy-kimball** [comment]
@knz, feel free to add any other person(s) who might want to review this. We're not sure who the right people/team would be.

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/server/tenant.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkd-LjsDo_ZAOpyv5Wq:-Mkdq9qY4uHo2gg4C-RI:b-uzi1um) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L334)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

after the `SampleCPU` change - not needed.
</blockquote></details>

How come it's not still needed? If we added code after this block, we don't want to execute it in this error case.

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/server/tenant.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkd-LjsDo_ZAOpyv5Wq:-Mke9Xko4LNKU5efQcsQ:b-68q1kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbe1213cef8dc691a8cf5eab910cb45b67fb2a/pkg/server/tenant.go#L334)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How come it's not still needed? If we added code after this block, we don't want to execute it in this error case.
</blockquote></details>

OK. Added a return.

**Review by andy-kimball — COMMENTED** [review]
:lgtm:, but make sure to get a review from someone on the Server team before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-:-MkeCbrT-y-ZQOUVNqHz:b-xcigwf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

**Review by knz — COMMENTED** [review]
you need to update the PR description with the new contents of the commit message.

> Reviewed 3 of 5 files at r1, 1 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-:-MkguILlDqhxVAJuuZYh:b-pi037y)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @darinpp)

**Review by darinpp — COMMENTED** [review]
done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-:-Mkh-TvV7EFXMCPl2d-k:bb7f2mf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @darinpp)

**Review by knz — APPROVED** [review]

**Comment by knz** [comment]
@dhartunian you may want to have a look too

**Comment by andy-kimball** [comment]
There's one potential problem I wanted to call out: because we're updating the server's instance of `RuntimeStatSampler`, which is shared with the regular `vars` endpoint, this will cause "de-sync'ing" of related metrics. For example, the instantaneous CPU won't "match" the CPU percentage. Generally, we want a reasonably consistent snapshot every 10 seconds, with each of the metrics roughly corresponding with one another by timestamp. By overwriting just some of these values every, say, 3 seconds, overlaying different metrics on top of one another by time won't work very well.

Perhaps we should use a different instance of `RuntimeStatSampler` for instant values so we don't mix with the background values we collect?

**Review by darinpp — COMMENTED** [review]
Good point. There is actually nothing that requires the two sets to share metrics. I changed the new endpoint to use separate gauges.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-:-MkhdWRc9gYtktdJBMdm:b-xx9ejo)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz)

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @darinpp and @knz)

___
*[pkg/server/tenant.go, line 321 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MklnGjv4Om8AlHdfVf9:-MklnGjv4Om8AlHdfVfA:blgc5cv) ([raw file](https://github.com/cockroachdb/cockroach/blob/07205b80faeb2bf9d5b3f1d9d17d41c891b99f67/pkg/server/tenant.go#L321)):*
> ```Go
> 
> 	return func(w http.ResponseWriter, r *http.Request) {
> 		if err := cpuTime.Get(os.Getpid()); err != nil {
> ```

We're back to not sharing this code, which I don't think is great. Can't you encapsulate this code in a little function, like we do with `GetUserCPUSeconds`, and then use it both here and from the sampler? Something like:
```
func GetCPUSeconds(ctx context.Context) (userTimeNs, sysTimeNs int64, err error)
```

I actually think we should move `GetUserCPUSeconds` to `tenant.go` and make it use the new `GetCPUSeconds` function. `GetUserCPUSeconds` is a very specialized function that shouldn't be in a shared file, IMO.

**Review by dhartunian — COMMENTED** [review]
:lgtm: I like Andy's suggestions re: code organization.

> Reviewed 2 of 5 files at r1, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-:-Mkm1Hx2BrzY3R7Gr6-R:br9k2yp)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @darinpp, @dhartunian, and @knz)

___
*[pkg/ccl/serverccl/tenant_vars_test.go, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkm1nNXB4gsOi96NK9i:-Mkm1nNXB4gsOi96NK9j:b-kh3aed) ([raw file](https://github.com/cockroachdb/cockroach/blob/07205b80faeb2bf9d5b3f1d9d17d41c891b99f67/pkg/ccl/serverccl/tenant_vars_test.go#L62)):*
> ```Go
> 		"invalid non-200 status code %v from tenant", resp.StatusCode)
> 
> 	prometheusMetricStringPattern := `^(?P<metric>\w+)(?:\{` +
> ```

Minor: We import the prometheus client library as a dependency so you might be able to use it to parse the metrics directly instead of manually via regex.

**Review by darinpp — COMMENTED** [review]
done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-:-MknzMjv2rOyhVcxiZnU:bb7f2mf)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @dhartunian, and @knz)

___
*[pkg/ccl/serverccl/tenant_vars_test.go, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-Mkm1nNXB4gsOi96NK9i:-MknzOcDB5NeElU9J7mN:b-ap20yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/07205b80faeb2bf9d5b3f1d9d17d41c891b99f67/pkg/ccl/serverccl/tenant_vars_test.go#L62)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Minor: We import the prometheus client library as a dependency so you might be able to use it to parse the metrics directly instead of manually via regex.
</blockquote></details>

switched to use Prometheus client
___
*[pkg/server/tenant.go, line 321 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MklnGjv4Om8AlHdfVf9:-MknzS_B4DH9b_eacmjA:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/07205b80faeb2bf9d5b3f1d9d17d41c891b99f67/pkg/server/tenant.go#L321)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We're back to not sharing this code, which I don't think is great. Can't you encapsulate this code in a little function, like we do with `GetUserCPUSeconds`, and then use it both here and from the sampler? Something like:
```
func GetCPUSeconds(ctx context.Context) (userTimeNs, sysTimeNs int64, err error)
```

I actually think we should move `GetUserCPUSeconds` to `tenant.go` and make it use the new `GetCPUSeconds` function. `GetUserCPUSeconds` is a very specialized function that shouldn't be in a shared file, IMO.
</blockquote></details>

Done

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @darinpp, @dhartunian, and @knz)

___
*[pkg/server/tenant.go, line 292 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkoYycWDZW4vwh36eEL:-MkoYycWDZW4vwh36eEM:b-9hw9dl) ([raw file](https://github.com/cockroachdb/cockroach/blob/90a2f8087a96545a600974e7b9adc535f26dc380/pkg/server/tenant.go#L292)):*
> ```Go
> 	getUserCPUSec := func(ctx context.Context) float64 {
> 		userTimeMS, _, err := status.GetCPUTime(ctx)
> 		if err != nil {
> ```

Don't need to log this error, as it's already logged in `GetCPUTime`.
___
*[pkg/server/status/runtime.go, line 450 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkoYls16wOVLNX_eJU7:-MkoYls16wOVLNX_eJU8:bbbhm1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/90a2f8087a96545a600974e7b9adc535f26dc380/pkg/server/status/runtime.go#L450)):*
> ```Go
> 	userTimeMS, sysTimeMS, err := GetCPUTime(ctx)
> 	if err != nil {
> 		log.Ops.Errorf(ctx, "unable to get cpu usage: %v", err)
> ```

This error has already been logged, you don't need to log again.
___
*[pkg/server/status/runtime.go, line 694 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkoYa2z8WR4igNvEGKw:-MkoYa2z8WR4igNvEGKx:b-s3xm65) ([raw file](https://github.com/cockroachdb/cockroach/blob/90a2f8087a96545a600974e7b9adc535f26dc380/pkg/server/status/runtime.go#L694)):*
> ```Go
> 
> // GetCPUTime returns the cumulative user/system time (in ms) since the process start.
> func GetCPUTime(ctx context.Context) (userTimeMS, sysTimeMS int64, err error) {
> ```

NIT: I think this should be `userTimeMs` rather than capitalizing as `MS`.

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @dhartunian, and @knz)

___
*[pkg/server/tenant.go, line 292 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkoYycWDZW4vwh36eEL:-MkrwWuZ1FShLtvZvM5j:b4g583z) ([raw file](https://github.com/cockroachdb/cockroach/blob/90a2f8087a96545a600974e7b9adc535f26dc380/pkg/server/tenant.go#L292)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Don't need to log this error, as it's already logged in `GetCPUTime`.
</blockquote></details>

I removed the logging in `GetCPUTime`. Leaving the log here as it is higher up in the call stack and will log the caller.
___
*[pkg/server/status/runtime.go, line 450 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkoYls16wOVLNX_eJU7:-MkrwlBZ-WYTqSkZvRv-:b4g583z) ([raw file](https://github.com/cockroachdb/cockroach/blob/90a2f8087a96545a600974e7b9adc535f26dc380/pkg/server/status/runtime.go#L450)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This error has already been logged, you don't need to log again.
</blockquote></details>

I removed the logging in `GetCPUTime`. Leaving the log here as it is higher up in the call stack and will log the caller.
___
*[pkg/server/status/runtime.go, line 694 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70750#-MkoYa2z8WR4igNvEGKw:-MkrwnR6BccRcsamYfVB:b3uucfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/90a2f8087a96545a600974e7b9adc535f26dc380/pkg/server/status/runtime.go#L694)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I think this should be `userTimeMs` rather than capitalizing as `MS`.
</blockquote></details>

Changed the abbreviations to mixed case.

**Review by andy-kimball — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70750)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @dhartunian and @knz)

**Pull request merged**
